### PR TITLE
Add support for IGD version 2

### PR DESCRIPTION
**Pull request by Chocobo1**
If possible, please cherry pick into RC_1_0 & RC_0_16 branch.

### CODE REVIEW

**Comment by arvidn** [comment]
would you mind extending the unit test to cover both versions as well? You could just add another parameter to run_upnp_test to specify version.

**Comment by Chocobo1** [comment]
> would you mind extending the unit test to cover both versions as well?

Sure, I'll ping you when I finish.

BTW, there are some errors when running `make check`, log attached:
```
$ make check -k
make  test_primitives test_recheck test_stat_cache test_file test_privacy test_priority test_auto_unchoke test_bdecode_performance test_checking test_fast_extension test_http_connection test_lsd test_metadata_extension test_pe_crypto test_pex test_read_piece test_resume test_ssl test_storage test_time_critical test_super_seeding test_swarm test_torrent test_tracker test_trackers_extension test_transfer enum_if test_utp test_session test_web_seed test_url_seed test_remap_files test_file_progress
make[1]: Entering directory '/home/asdf/Desktop/libtorrent-git/libtorrent/test'
  CXXLD    test_primitives
../src/.libs/libtorrent-rasterbar.so: undefined reference to `boost::random::random_device::operator()()'
../src/.libs/libtorrent-rasterbar.so: undefined reference to `boost::random::random_device::~random_device()'
../src/.libs/libtorrent-rasterbar.so: undefined reference to `boost::random::random_device::random_device()'
collect2: error: ld returned 1 exit status
Makefile:1095: recipe for target 'test_primitives' failed
make[1]: *** [test_primitives] Error 1
```
Running on up-to-date Arch Linux, boost package is from Arch repo (fetched via pacman)

**Comment by Chocobo1** [comment]
I added more commits and I think it is ready now, take a look. :)

File: src/upnp.cpp [hunk 1]
```diff
@@ -965,8 +965,31 @@ void upnp::on_upnp_xml(error_code const& e
 		return;
 	}
 
+	// determine IGD version
+	std::string igd_str(p.get_body().begin, p.get_body().end);
```
Comment (arvidn) [inline src/upnp.cpp]: looking at this code, doesn't it look a bit backwards? There is an xml parser below, so why just do a string-match search first? It seems more natural to parse out the version after the xml parser has found the key. I realize it's not very likely the xml document will have an occurrence of this string elsewhere, but the code implies that it is impossible, and it is slightly confusing because of it, in my opinion.

File: src/upnp.cpp [hunk 2]
```diff
@@ -965,8 +965,31 @@ void upnp::on_upnp_xml(error_code const& e
...
+		snprintf(msg, sizeof(msg), "invaild IGD version from %s, falling back to version 1"
+			, d.url.c_str());
+		log(msg, l);
+		igd_ver = 1;
+	}
+	d.igd_version = igd_ver;
+
 	parse_state s;
-	s.reset("urn:schemas-upnp-org:service:WANIPConnection:1");
+	static std::string wanip_service = std::string("urn:schemas-upnp-org:service:WANIPConnection:") + to_string(d.igd_version).elems;
```
Comment (arvidn) [inline src/upnp.cpp]: I think it would make more sense find_control_url to recognize the string prefix and parse out the version itself. It may even be a good opportunity to merge looking for WANIPConnection and PPPConnection as well.

Currently, there are just two lookups, if the WANIPConnection can't be found, we search for WANPPPConnection instead. See: https://github.com/arvidn/libtorrent/pull/23/files#diff-54a4016ca61d648addae2913638804d9R1004

Even if those two paths aren't merged, doesn't PPPConnection have to support version 2 as well?

File: test/test_upnp.cpp [hunk 3]
```diff
@@ -46,17 +46,27 @@ using namespace libtorrent;
...
+	"<s:Body><u:AddPortMapping xmlns:u=\"urn:schemas-upnp-org:service:WANIPConnection:2\">"
+	"</u:AddPortMapping></s:Body></s:Envelope>";
 
-char soap_delete_response[] =
+char soap_delete_response_v1[] =
 	"<s:Envelope xmlns:s=\"http://schemas.xmlsoap.org/soap/envelope/\" "
 	"s:encodingStyle=\"http://schemas.xmlsoap.org/soap/encoding/\">"
 	"<s:Body><u:DeletePortMapping xmlns:u=\"urn:schemas-upnp-org:service:WANIPConnection:1\">"
 	"</u:DeletePortMapping></s:Body></s:Envelope>";
+char soap_delete_response_v2[] =
```
Comment (arvidn) [inline test/test_upnp.cpp]: if this was an array of two strings...

File: test/test_upnp.cpp [hunk 4]
```diff
@@ -141,7 +151,10 @@ void run_upnp_test(char const* root_filename, char const* router_model, char con
 	fclose(xml_file);
 
 	std::ofstream xml(control_name, std::ios::trunc);
-	xml.write(soap_add_response, sizeof(soap_add_response)-1);
+	if(igd_version == 1)
```
Comment (arvidn) [inline test/test_upnp.cpp]: all of these if-conditions could just be an array lookup with igd_version as the index

File: test/test_upnp.cpp (hunk 3 above)
Comment (Chocobo1) [inline test/test_upnp.cpp]: Fixed.

File: test/test_upnp.cpp (hunk 4 above)
Comment (Chocobo1) [inline test/test_upnp.cpp]: Fixed.

File: src/upnp.cpp (hunk 1 above)
Comment (Chocobo1) [inline src/upnp.cpp]: > looking at this code, doesn't it look a bit backwards? There is an xml parser below, so why just do a string-match search first?

I do see there is a xml parser down there, yet I am reluctant to use it, because either I need to modify `find_control_url` (which is already doing a lot of work and hard to hack in there), or I need to create another callback for this simple job.

I am not sure what you suggest, but it seems writing another cb is inevitable !?

File: src/upnp.cpp (hunk 2 above)
Comment (Chocobo1) [inline src/upnp.cpp]: > I think it would make more sense find_control_url to recognize the string prefix and parse out the version itself. It may even be a good opportunity to merge looking for WANIPConnection and PPPConnection as well.

I don't see any easy way to achieve this, unless a rewrite of `find_control_url`, what do you suggest to do that?

> Even if those two paths aren't merged, doesn't PPPConnection have to support version 2 as well?

No, from [InternetGatewayDevice:2  Device Template Version 1.01](http://upnp.org/specs/gw/UPnP-gw-InternetGatewayDevice-v2-Device.pdf),
```
1.2. Focus and Goals for DCP version 2.0 
Note: Due to lack of interest,  WANPPPConnection service was NOT updated to reflect the changes made to WANIPConnection. 
```
so `WANPPPConnection` only have version 1.

**Comment by Chocobo1** [comment]
Nevermind, I did another approach in c5f7210, pls take a look.

File: test/setup_transfer.cpp [hunk 5]
```diff
@@ -869,7 +869,7 @@ int start_web_server(bool ssl, bool chunked_encoding, bool keepalive)
 	} while (ec);
 
 	char buf[200];
-	snprintf(buf, sizeof(buf), "python ../web_server.py %d %d %d %d"
+	snprintf(buf, sizeof(buf), "python2 ../web_server.py %d %d %d %d"
```
Comment (arvidn) [inline test/setup_transfer.cpp]: hm.. I'm not sure about this. could finding the python executable be put in the python script perhaps? I take it your system doesn't have "python" in the path? (or is the problem that web_server.py is not python3 compatible?).

perhaps the python scripts should all start with "#!/usr/bin/env python". Fixing the python invocations could perhaps be a separate patch though

File: test/test_upnp.cpp [hunk 6]
```diff
@@ -46,17 +46,25 @@ using namespace libtorrent;
 broadcast_socket* sock = 0;
 int g_port = 0;
 
-char soap_add_response[] =
+char soap_add_response[][270] = {
```
Comment (arvidn) [inline test/test_upnp.cpp]: I believe you could also say: char const* soap_add_request[] = 
to not have to specify the length of the strings

File: src/upnp.cpp [hunk 7]
```diff
@@ -898,13 +889,18 @@ TORRENT_EXTRA_EXPORT void find_control_url(int type, char const* string, parse_s
 	else if (type == xml_string)
 	{
 		if (state.tag_stack.empty()) return;
-//		std::cout << " " << string << std::endl;
+//		std::cout << " " << string << std::endl;}
 		if (!state.in_service && state.top_tags("service", "servicetype"))
 		{
-			if (string_equal_no_case(string, state.service_type))
+			if (string_equal_no_case(string, "urn:schemas-upnp-org:service:WANIPConnection:1")
```
Comment (arvidn) [inline src/upnp.cpp]: this is a great change. thanks! Do you think there's a simple way of creating a bias towards version 2? or does that even make sense? The case I'm thinking of is if a router provides multiple services, right now we would pick the first one, previously we would prefer the WANIPConnection over PPPConnection.

File: test/setup_transfer.cpp (hunk 5 above)
Comment (Chocobo1) [inline test/setup_transfer.cpp]: > or is the problem that web_server.py is not python3 compatible?

This one, calling `python` refers to `python3` when python3 is installed.

> perhaps the python scripts should all start with "#!/usr/bin/env python"

That won't help on this problem. Nevertheless, that should be added.

>Fixing the python invocations could perhaps be a separate patch though

Then I'll drop it in this PR.

File: test/test_upnp.cpp (hunk 6 above)
Comment (Chocobo1) [inline test/test_upnp.cpp]: You're right, thanks!

File: src/upnp.cpp (hunk 7 above)
Comment (Chocobo1) [inline src/upnp.cpp]: > this is a great change. thanks! Do you think there's a simple way of creating a bias towards version 2? or does that even make sense?

Hmmm, I think there is no special need to favor version 2. Just supporting it is enough.

> The case I'm thinking of is if a router provides multiple services, right now we would pick the first one, previously we would prefer the WANIPConnection over PPPConnection.

I don't have much knowledge on this, but I think it will work fine. `WANIPConnection` is likely to be implemented on all IGDs:
```
2.2. Device Model Table 1
WANIPConnection
Required for modems that support IPv4 based connections
```

**Comment by Chocobo1** [comment]
OK, I squashed the commits.
